### PR TITLE
HIVE-27526: Cleaner should honor compaction writeIdHwm

### PR DESCRIPTION
**Pull request by deniskuzZ**
### What changes were proposed in this pull request?

Cleaner should honor compaction writeIdHwm. Introduced ValidCleanerWriteIdList

### Why are the changes needed?

Cleaner check remaining logic was inaccurate

### Does this PR introduce _any_ user-facing change?

No
### Is the change a dependency upgrade?

No

### How was this patch tested?

UT

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnUtils.java [hunk 1]
```diff
@@ -77,15 +77,14 @@ public static ValidTxnList createValidTxnListForCleaner(GetOpenTxnsResponse txns
...
       if (abortedBits.get(i)) {
-        exceptions[i] = txnId;
+        exceptions[i++] = txnId;
       } else {
         if (isAbortCleanup) {
-          exceptions[i] = txnId;
+          exceptions[i++] = txnId;
         } else {
-          assert false : JavaUtils.txnIdToString(txnId) + " is open and <= hwm:" + highWatermark;
+          assert (minOpenTxn == Long.MAX_VALUE) : JavaUtils.txnIdToString(txnId) + " is open and <= hwm:" + highWatermark;
```
Comment (SourabhBadhya) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnUtils.java:85]: I think the fundamental idea behind keeping it as false, is that compaction can't treat writes from an open txn if they are present within the highWatermark. Won't this condition add an open txn to the list which is later used by compaction?

Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnUtils.java:85]: - Open txns are not added to the exceptions list when isAbortCleanup=false, only aborts.
- When there are no open write txns and we rely on MinHistoryWriteId, minOpenTxn=Long.MAX_VALUE. 
- `assert` is only executed during the test phase, unless -ea option is provided

Comment (SourabhBadhya) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnUtils.java:85]: AFAIK, if an open txn is present and it is not added to the exceptions list or not failed. Then it is treated as a valid txn when fetching the associated write IDs.

Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnUtils.java:85]: there shouldn't be any open txns below the highWatermark, Cleaner only checks the committed or aborted writes
````
      return new ValidReadTxnList(exceptions, bitSet, highWatermark, Long.MAX_VALUE);
    } else {
      return new ValidReadTxnList(exceptions, abortedBits, highWatermark, Long.MAX_VALUE);
````
it's set to Long.MAX_VALUE

Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnUtils.java:85]: i think we need to cap hwm by NEXT_WRITE_ID-1, WDYT?

Comment (SourabhBadhya) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnUtils.java]: Yes, now it makes sense.

File: storage-api/src/java/org/apache/hadoop/hive/common/ValidCleanerWriteIdList.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.common;
+
+import java.util.BitSet;
+import java.util.Optional;
+
+public class ValidCleanerWriteIdList extends ValidReaderWriteIdList {
```
Comment (veghlaci05) [inline storage-api/src/java/org/apache/hadoop/hive/common/ValidCleanerWriteIdList.java:24]: Please add a short class description about the purpose of this class (different range validation)

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/handler/TaskHandler.java [hunk 3]
```diff
@@ -121,7 +122,8 @@ protected ValidReaderWriteIdList getValidCleanerWriteIdList(CompactionInfo info,
     // been some delta/base dirs
     assert rsp != null && rsp.getTblValidWriteIdsSize() == 1;
 
-    return TxnCommonUtils.createValidReaderWriteIdList(rsp.getTblValidWriteIds().get(0));
+    return new ValidCleanerWriteIdList(
+        TxnCommonUtils.createValidReaderWriteIdList(rsp.getTblValidWriteIds().get(0)));
```
Comment (veghlaci05) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/handler/TaskHandler.java:126]: What about creating a separate method like createValidCleanerWriteIdList(..)?

File: ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCleaner.java [hunk 4]
```diff
@@ -1117,6 +1119,61 @@ public void testReady() throws Exception {
...
+    ctxnid = compactInTxn(rqst);
+    addDeltaFile(t, p, 20, 24, 5, ctxnid);
+
+    startCleaner();
+    txnHandler.abortTxn(new AbortTxnRequest(openTxnId));
+
+    ShowCompactResponse rsp = txnHandler.showCompact(new ShowCompactRequest());
+    Assert.assertEquals(2, rsp.getCompactsSize());
+    Assert.assertEquals(TxnStore.SUCCEEDED_RESPONSE, rsp.getCompacts().get(0).getState());
+    Assert.assertEquals(TxnStore.CLEANING_RESPONSE, rsp.getCompacts().get(1).getState());
```
Comment (veghlaci05) [inline ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCleaner.java:1161]: The result of showCompact is not ordered at all. Wouldn't be it flaky?

**Review by veghlaci05 — CHANGES_REQUESTED** [review]

File: ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCleaner.java (hunk 4 above)
Comment (deniskuzZ) [reply ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCleaner.java:1161]: it is ordered by default with TxnQueries.SHOW_COMPACTION_ORDERBY_CLAUSE

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/handler/TaskHandler.java (hunk 3 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/handler/TaskHandler.java:126]: it's just used in 1 single place, i would prefer not to create 1 more util method

File: storage-api/src/java/org/apache/hadoop/hive/common/ValidCleanerWriteIdList.java (hunk 2 above)
Comment (deniskuzZ) [reply storage-api/src/java/org/apache/hadoop/hive/common/ValidCleanerWriteIdList.java:29]: added javadoc

**Review by veghlaci05 — APPROVED** [review]
LGTM, +1

**Pull request merged**